src/backends/interflop-mca/interflop_mca.c

```diff
@@ -122,135 +152,85 @@ static double _mca_rand(void) {
-  ix = hex & DOUBLE_ERASE_SIGN;
-  // shift exponent to have LSB on position 0 and complement
-  exp = (ix >> DOUBLE_PMAN_SIZE) - DOUBLE_EXP_COMP;
-  return exp;
-}
+/* noise = rand * 2^(exp) */
+__float128 qnoise(int exp) {
+  /* random number in (-0.5, 0.5) */
+  const binary64 brand = {.f64 = _mca_rand() - 0.5};
+  const int32_t brand_exp = brand.ieee.exponent - DOUBLE_EXP_COMP;
+  const int32_t noise_exp = brand_exp + exp;
 
-/* Returns the MCA noise for the quad format
- * qnoise = 2^(exp)*d_rand */
-static inline __float128 qnoise(int exp) {
-  double d_rand = (_mca_rand() - 0.5);
-  uint64_t u_rand = *((uint64_t *)&d_rand);
-  __float128 noise;
-  uint64_t hx, lx;
+  binary128 noise;
 
-  if (exp > QUAD_EXP_MAX) { /*exceed max exponent*/
-    SET_FLT128_WORDS64(noise, QINF_hx, QINF_lx);
-    return noise;
+  /* specials */
+  if (exp == 0) {
+    noise.f128 = brand.f64;
   }
-  if (exp < -QUAD_EXP_MIN) { /*subnormal*/
-    // test for minus infinity
-    if (exp <= -(QUAD_EXP_MIN + QUAD_PMAN_SIZE)) {
-      SET_FLT128_WORDS64(noise, QMINF_hx, QMINF_lx);
-      return noise;
-    } else{// noise will be a subnormal
-    // build HX with sign of d_rand, exp
-    uint64_t u_hx = ((uint64_t)(-QUAD_EXP_MIN + QUAD_EXP_COMP))
-                    << QUAD_HX_PMAN_SIZE;
-    // add the sign bit
-    uint64_t sign = u_rand & DOUBLE_GET_SIGN;
-    u_hx = u_hx + sign;
-    // erase the sign bit from u_rand
-    u_rand = u_rand << 1;
-
-    uint64_t u_lx=0;
-    int relative_exp=-exp - QUAD_EXP_MIN;
-    if (relative_exp< QUAD_HX_PMAN_SIZE) {
-      // the higher part of the noise start in HX of noise
-      // set the mantissa part: U_rand>> by -exp-QUAD_EXP_MIN
-      u_hx += u_rand >> (relative_exp + QUAD_EXP_SIZE + 1 /*SIGN_SIZE*/);
-      // build LX with the remaining bits of the noise
-      // (-exp-QUAD_EXP_MIN-QUAD_HX_PMAN_SIZE) at the msb of LX
-      // remove the bit already used in hx and put the remaining at msb of LX
-      u_lx = u_rand << (QUAD_HX_PMAN_SIZE - relative_exp);
-      SET_FLT128_WORDS64(noise, u_hx, u_lx);
-    } else{
-      int32_t shift=relative_exp - QUAD_HX_PMAN_SIZE;
-      //shift by 64 and more is undifinied in the C norm
-      u_lx = u_rand >> shift;
-      SET_FLT128_WORDS64(noise, u_hx, u_lx);
-    }}
-    // char buf[128];
-    // int len=quadmath_snprintf (buf, sizeof(buf), "%+-#*.20Qe", width, noise);
-    // if ((size_t) len < sizeof(buf))
-    // printf ("subnormal noise %s\n", buf);
-    return noise;
+  /* We exceed the maximal exponent */
+  /* we return an sign(rand) * Infinity */
+  else if (noise_exp > QUAD_EXP_MAX) {
+    noise.ieee128.sign = brand.ieee.sign;
+    noise.ieee128.exponent = QUAD_EXP_MAX;
+    noise.ieee128.mantissa = 0;
+  }
+  /* We are in the subnormal case */
+  else if (noise_exp < -QUAD_EXP_MIN) {
+    /* test if result is below the minimal subnormal quad number */
+    if (noise_exp < -(QUAD_EXP_MIN + QUAD_PMAN_SIZE)) {
```

Comment:
Actually, we are never to be going in this case since the exponent value comes from a double and will never reach the QUAD_EXP_MIN bound. Moreover, your original code was wrong since this test tests if we are in the subnormal case and you set to -infinity after that.